lib/collection/src/wal_delta.rs

```diff
@@ -279,6 +280,119 @@ mod tests {
+        // all `0`s.
+        let mut a_clock_set = ClockSet::new();
+        let mut a_clock_0 = a_clock_set.get_clock();
+
+        // Let's create another operation on peer A and write it on all nodes:
+
+        // Create operation on peer A
+        let clock_tag = ClockTag::new(1, a_clock_0.id(), a_clock_0.tick_once());
+        let operation = OperationWithClockTag::new(mock_operation(4), Some(clock_tag));
+
+        // Write operation on peers A and B
+        for wal in [&a_wal, &b_wal] {
+            let mut operation = operation.clone();
+            let (_, _) = wal.lock_and_write(&mut operation).await.unwrap();
+            a_clock_0.advance_to(operation.clock_tag.unwrap().clock_tick);
+        }
+
+        // Then write operation on peer C from the forward proxy *first*
+        {
+            let mut operation = operation.clone();
+            let (_, _) = c_wal.lock_and_write(&mut operation).await.unwrap();
+            a_clock_0.advance_to(operation.clock_tag.unwrap().clock_tick);
+        }
+
+        // And now let's assume that operation, that was sent from node A to C directly was written
+        // to WAL, but the response, that should have advanced the clock set on node A failed.
+        //
+        // Currently, we simply ignore all errors from the nodes in `Partial` state, so it won't
+        // mark node C as `Dead`.
+        {
+            let mut operation = operation.clone();
+            let (_, _) = c_wal.lock_and_write(&mut operation).await.unwrap();
+            // we do not advance `a_clock_0` here
+        }
+
+        // Now, if we assume that:
+        // - we increment local clock map, when we receive an operation with `clock_tick = 0`
+        // - and we reject older operations
+        //   - (and we have to do so, otherwise we can't guarantee that our WAL has monotoncally increasing versions)
+        //
+        // Then the next operation will be *rejected* on node C, and C will be permanently missing
+        // `mock_operation(5)`.
+        //
+        // I think, this *may* be (fully? partially?) prevented by the logic to set "cutoff" clocks
+        // after stream records transfer, but I'm not 100% sure. :/
+
+        // Create operation on peer A
+        let clock_tag = ClockTag::new(1, a_clock_0.id(), a_clock_0.tick_once());
+        let operation = OperationWithClockTag::new(mock_operation(5), Some(clock_tag));
+
+        // Write operation on peers A and B
+        for wal in [&a_wal, &b_wal] {
+            let mut operation = operation.clone();
+            let (_, _) = wal.lock_and_write(&mut operation).await.unwrap();
+            a_clock_0.advance_to(operation.clock_tag.unwrap().clock_tick);
+        }
+
+        // Then try to write operation on peer C
+        {
+            let mut operation = operation.clone();
+            // This should be rejected, but we currently accept all operations, so I can't assert this *yet* 🤷‍♀️
```

Comment:
Do we also want to reject operations with the same clock tick? Because if we do not, I don't think we have this problem. Then you'd just get operation 5 (or whichever it is) a second time. We do not have more than one forward proxy to a shard at a time.

Reply:
If we do not reject current tick, then we'll just start having multiple operations with the same tick again. And in that case, what's the point of bothering with incrementing clock map in the first place then?

Reply:
Disregard that. I see that you might have a point here. In this case we *kinda* don't care about repetitions, and can just return the last operation with requred tick. Hmmmmm.......

Reply:
Isn't that a compromise good enough? (rejecting lower, allowing the same)

Can we think of other scenarios where this incrementing may be problematic?